tests/distributed/test_custom_all_reduce.py

```diff
@@ -18,17 +20,24 @@
 
 
 @ray.remote(num_gpus=1, max_calls=1)
-def graph_allreduce(world_size, rank, distributed_init_port):
+def graph_allreduce(tp_size, pp_size, rank, distributed_init_port):
     del os.environ["CUDA_VISIBLE_DEVICES"]
     device = torch.device(f"cuda:{rank}")
     torch.cuda.set_device(device)
-    init_test_distributed_environment(1, world_size, rank,
+    init_test_distributed_environment(pp_size, tp_size, rank,
                                       distributed_init_port)
 
-    custom_all_reduce.init_custom_ar()
+    group = get_tensor_model_parallel_group()
+    # we use the first group to communicate once
+    # and the second group to communicate twice
+    # and so on
+    # this is used to demonstrate that each group can
+    # communicate independently
+    num_communication = rank // tp_size + 1
+
     for sz in test_sizes:
         for dtype in [torch.float32, torch.float16, torch.bfloat16]:
```

Comment:
nit: Maybe we can use `dtype` as an input parameter so that the dtypes can be tested separately?

Reply:
Quite difficult. The input of `test_custom_allreduce`, is coupled with the input of `multi_process_tensor_parallel`, which is used elsewhere. In other words, we cannot modify the input parameter inside just `tests/distributed/test_custom_all_reduce.py` 🤣